program/rust/src/processor/upd_price.rs

```diff
@@ -127,27 +127,74 @@ pub fn upd_price(
     // Check clock
     let clock = Clock::from_account_info(clock_account)?;
 
-    let mut publisher_index: usize = 0;
+    let publisher_index: usize;
     let latest_aggregate_price: PriceInfo;
 
     // The price_data borrow happens in a scope because it must be
     // dropped before we borrow again as raw data pointer for the C
     // aggregation logic.
     {
         // Verify that symbol account is initialized
-        let price_data = load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+        let mut price_data = load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
 
-        // Verify that publisher is authorized
-        while publisher_index < try_convert::<u32, usize>(price_data.num_)? {
-            if price_data.comp_[publisher_index].pub_ == *funding_account.key {
-                break;
+        // Verify that publisher is authorized by initially binary searching
+        // for the publisher's component in the price account. The binary
+        // search might not work if the publisher list is not sorted; therefore
+        // we fall back to a linear search and during the search we try to bring
+        // the publishers list closer to a sorted state. (We do not sort them
+        // entirely to avoid exceeding the compute budget.)
+        let mut binary_search_result = None;
+
+        {
+            // Binary search to find the publisher key. Rust std binary search is not used because
+            // they guarantee valid outcome only if the array is sorted whereas we want to rely on
+            // a Equal match if it is a result on an unsorted array. Currently the rust
+            // implementation behaves the same but we do not want to rely on api internals.
+            let mut left = 0;
```

Comment:
I hesitated to do this because i don't know whether function call adds cu overhead. Let me do it with `inline` and if it's fine I'll add the test. That being said, in our test since all the publishers are publishing prices the binary search gets test for each publisher. I'll add a comment explaining it too.